README.md

```diff
@@ -1,62 +1,81 @@
-# arrow
+# Arrow
 
-“... even if the previous millisecond is closer to us than the birth of the universe, it is equally out of reach.” 
+“... even if the previous millisecond is closer to us than the birth of the universe, it is equally out of reach.”
 ― Jean-Christophe Valtat, Luminous Chaos
 
-## concept
+## Concept
 
-This library implements a generalized version of the [Gillespie Algorithm](https://en.wikipedia.org/wiki/Gillespie_algorithm), a stochastic approach to numerically solving discrete systems. Each iteration, the algorithm will calculate the propensities for each reaction given a rate and the counts of the reactants present in the current state of the system, then selects one reaction to occur and the interval of time between the previous reaction and the current reaction. Iterating this produces a trajectory (or `history`) of the state vector over the course of the simulation.
+This library implements a generalized version of the [Gillespie
+Algorithm](https://en.wikipedia.org/wiki/Gillespie_algorithm), a stochastic
+approach to numerically solving discrete systems. Each iteration, the algorithm
+will calculate the propensities for each reaction given a rate and the counts
+of the reactants present in the current state of the system, then selects one
+reaction to occur and the interval of time between the previous reaction and
+the current reaction. Iterating this produces a trajectory (or `history`) of
+the state vector over the course of the simulation.
 
-## installation
+## Installation
 
-Add the following to your `requirements.txt`, or `pip install stochastic-arrow`:
+Add the following to your `requirements.txt`, or
+`pip install stochastic-arrow`:
 
     stochastic-arrow==0.0.1
 
-## usage
+## Usage
 
-The `arrow` library presents a single class as an interface, `StochasticSystem`, which operates on a set of reactions (encoded as a `numpy` matrix) and associated reaction rates:
+The `arrow` library presents a single class as an interface,
+`StochasticSystem`, which operates on a set of reactions (encoded as a `numpy`
+matrix of stoichiometrix coefficients) and associated reaction rates:
 
 ```python
 from arrow import StochasticSystem
 import numpy as np
 
-# Each row is a reaction and each column is a molecular species (or other entity).
-# The top reaction here means that the first and second elements combine to create the third,
-# while the fourth is unaffected.
-reactions = np.array([
-    [-1, -1, 1, 0],
-    [-2, 0, 0, 1],
-    [1, 1, -1, 0]])
+# Each column is a reaction and each row is a molecular species (or other
+# entity). The first reaction here means that the first and second elements
+# combine to create the third, while the fourth is unaffected.
+stoichiometry = np.array([
+    [-1, -2, +1],
+    [-1,  0, +1],
+    [+1,  0, -1],
+    [ 0, +1,  0]
```

Comment:
I gotta say, this transposed form is unpleasant to me, though if it is some kind of standard I understand. But having each reaction split out among every row upsets some organizational principle deep in my soul.

Reply:
There are a few different things colliding here.  In biochemical dynamics literature we write `dc/dt = Sv` where `Sv` is the matrix-vector product between the stoichiometric matrix and a vector of reaction rates; in this form, `S` has to have reactions along the columns and metabolites along the rows.  This is how `S` is defined in the FBA literature, too.

However, there's an issue - we don't just use the stoichiometry to decide what a reaction does, we also use it to decide the rate of reaction (`v`), since that is dependent upon reactant abundances.  So stoichiometry is actually serving two different purposes; first, to calculate rates (our propensities), and second, to decide what the update actually is, given that an event occurs (this is the true meaning of stoichiometry).

On top of all that, we're naturally doing things on a per-reaction basis, since Gillespie's SSA is designed such that only one reaction occurs at each step.  Thus we don't really get much out of the standard dynamic form.

So, in short: this *is* the standard form of a stoichiometric matrix, even if it's inconvenient and inefficient for this particular problem.  I think we should drop the matrix representation anyway (#8), since in practice these systems are so sparse (#9), at which point this becomes a non-issue.  I can prioritize that change.

Reply:
Awesome, thanks for the explanation. So this is a classic mismatch between theoretical forms and what makes sense in a computer. I would say we can present the interface in the dynamic way that everyone expects, and do the computation with the transpose which makes more sense for the computer to do. Retrieving the reaction stoichiometry with a simple index into the whole matrix seems too natural not to use, but yeah the sparse representation would make all of this moot anyway.

Reply:
What's really odd is that I've tried switching to a Fortran memory ordering, and if anything the evaluation time goes up.  It's hard to say because the variability in each evaluation is on the order of maybe 20% of the average run time.  I'm going to leave it alone for now.